### ISSUE TITLE
proposed NSRegularExpression extensions

### ISSUE THREAD

**Issue by gurgeous** [SwifterSwift/SwifterSwift#896]
I figured I'd open this up for discussion before putting together a PR. These are from my current project. What do you guys think?

### NSRegularExpression

```swift
// fatalError when bad pattern, useful for statics and other hardcoded regex
init(_ pattern: String, options: NSRegularExpression.Options = [])

// return true if the string matches
hasMatch(_ string: String) -> Bool

// run on entire string, annoying NSRange no longer required
enumerateMatches
firstMatch
matches
numberOfMatches
rangeOfFirstMatch
replaceMatches
stringByReplacingMatches
```

### String

```swift
// like matches(pattern)
matches(regex: NSRegularExpression) -> Bool

// regex.stringByReplacingMatches(self)
stringByReplacingMatches(regex: NSRegularExpression, withTemplate template: String) -> String
stringByReplacingMatches(pattern: String, withTemplate template: String) -> String
```

**Comment by guykogus** [comment]
The initialiser isn't necessary - you can just use `try!` with the standard init.

I already created versions of each `NSRegularExpression` function that uses `Range<String.Index>` instead of `NSRange`. It would be nice to have no range value at all. As an option, instead of creating duplicates of each function, we can change the `range` parameters to be optional and use the string's full range in case they're `nil`.

For `String`  we already have `matches(pattern:)` and `~=`. You could add new `replacingOccurrences` functions (to maintain naming) such as
```Swift
func replacingOccurrences(of regex: NSRegularExpression, with template: String, range searchRange: Range<String.Index>) -> String
```

**Comment by gurgeous** [comment]
> The initialiser isn't necessary - you can just use `try!` with the standard init.

This is annoying with swiftlint, but I don't feel strongly about it.

> I already created versions of each `NSRegularExpression` function that uses `Range<String.Index>` instead of `NSRange`. It would be nice to have no range value at all. As an option, instead of creating duplicates of each function, we can change the `range` parameters to be optional and use the string's full range in case they're `nil`.

👍 


> For `String` we already have `matches(pattern:)` and `~=`.

Hm. It's nice when these methods work with both NSRegularExpression and string pattern. The patterns are handy, but I often use the regex flavors for performance. That would be four methods total:

```swift
matches(regex)
matches(pattern) (we already have this)
~= regex
~= pattern (we already have this)
```


> You could add new `replacingOccurrences` functions (to maintain naming) such as
> 
> ```swift
> func replacingOccurrences(of regex: NSRegularExpression, with template: String, range searchRange: Range<String.Index>) -> String
> ```

sounds good

**Comment by guykogus** [comment]
I'll do the `NSRegularExpression` changes if you want to do the `String` changes.

**Comment by gurgeous** [comment]
Sounds good to me!

**Comment by LucianoPAlmeida** [comment]
I think this was resolved by #901 so closing ... if not feel free to reopen @gurgeous :)
